### ISSUE TITLE
Light on mesh

### ISSUE THREAD

**Issue by JPFrancoia** [vispy/vispy#1479]
Hi,

I already commented in https://github.com/vispy/vispy/issues/553, but I'm not looking for a full solution, just a quick fix. I actually have an idea, but I need to know how Vispy works before I can implement it.

I have a problem with lighting. Here is an object visualized with vispy:

![screen](https://user-images.githubusercontent.com/2399090/39407323-27953ecc-4bbc-11e8-8365-b3b550614f33.png)

But when I rotate it, it seems the lighting rotates as well:

![screen_dark](https://user-images.githubusercontent.com/2399090/39407332-41b6543a-4bbc-11e8-92e7-f2f23802df2f.png)

To visualize this object, I use an arcball camera. My question is: when is the lighting calculated/rendered ? When the mesh is loaded ? When the visual is created ? Or when the camera object is created ?

```python
mdata = geometry.MeshData(mesh.vertices, mesh.faces)
mesh = visuals.Mesh(meshdata=mdata, shading='flat', parent=self.view.scene)
```

So basically, which line is responsible for creating the lighting ?

Cheers

**Comment by djhoese** [comment]
Technically the lighting is calculated (the normals, etc) in the `Mesh` object, but they are supposed to be updated every time a visual component of the Mesh changes (see `_update_data` in `MeshVisual`). My guess, which I think you've alluded to, is that the light moves around with the object. That is, the camera is the only thing moving and the light and mesh object are in one place. You want the light and camera to stay in one place, but move the object. Or at least that's the end result.

My guess is that if you could trigger an `_update_data` after you have moved with the camera that it would update the lighting properly. Or it's possible the light is always supposed to stay in the same direction as the object which seems more likely the more I look at the code.

**Comment by JPFrancoia** [comment]
Thank you for your answer, it helps. I tried the ```_update_data``` solution, it didn't work. I also tried to delete and re-calculate my mesh, no success.

But what you said makes sense, I think the object and the lighting are fixed and the camera moves around (which actually makes sense...it's a camera object). It's not really intuitive since clicking and moving shows the object rotating, but it's actually not moving.

So I tried this little code:

```python
    def on_mouse_release(self, event):

        # new_meshes = dict()
        import random

        for name, mesh in self.meshes.items():
            # mesh._update_data()
            print(mesh.light_dir)
            mesh.light_dir = (random.randrange(0, 10), random.randrange(0, 10), random.randrange(0, 10))
```

When the mouse button is released, it randomly changes the ```light_dir``` of the object. And it "works", the light moves around the object and shine on a different angle.

So the workaround to my problem could be very simple, we just need to find the new coordinates of the camera after each movement, and recalculate the direction of the light.

**Comment by djhoese** [comment]
You may be able to edit the shader code of the mesh to get this work and to get it to be parameterized (based of a property on the mesh). I think @campagnola mentioned recently that there is another object or example that does something like this. That is, linking up with the camera to figure out where the camera is. Although...he may have been talking about lights ;) .

**Comment by JPFrancoia** [comment]
Hum maybe, it seems complicated.
Is there anything in the camera modules that give the direction of a camera ?

**Comment by JPFrancoia** [comment]
Or the transformation a camera is about to apply when the mouse button is released ?

**Comment by djhoese** [comment]
@JPFrancoia There may be some of this information in the camera class itself. I believe there are also transform_changed (not sure on exact name) events that you can connect to, but I'm not sure how this works off the top of my head.

**Comment by asnt** [comment]
@JPFrancoia As you found out, the light direction is determined by mesh.light_dir in the Mesh object. I have an implementation of a light that follows the camera direction in a pending PR https://github.com/asnt/vispy/blob/c09ff83cf7c8148ceb539a158881411a6ae4a68f/examples/basics/scene/mesh_texture.py#L35 (lines 35-48).

**Comment by JPFrancoia** [comment]
@asnt haaaaaa ! thank you very much. I'll try it very soon, it's crucial for my software.

Can you explain me how it works please ? I know the transform matrix of the camera is accessible (it's an attribute). I tried the dot product of this matrix with the light_dir vector to calculate the new light_dir.

I think you're doing something similar, but what do ```map``` and ```imap``  do ?

**Comment by JPFrancoia** [comment]
@asnt I tried you code. It indeed works when I rotate the camera. However, when I translate the camera (Shift + click, then move the mouse), it stops working. If I try to rotate the camera after translating it, it doesn't work (the light doesn't follow the camera anymore).

**Comment by JPFrancoia** [comment]
Perfect, it now works even after translation.

Thank you very much, I spent three days trying to find a fix for this issue. My lack of knowledge in maths...

**Comment by djhoese** [comment]
@asnt Thanks for your help on this. Does this need to be updated in your PR?

I also suppose the concatenate call could be moved to the initial_light_dir creation, right? Is there any use case where this component shouldn't be ignored?

**Comment by asnt** [comment]
@djhoese You are welcome.
Yes, it should be updated. Feel free to do it, if you feel like right now. I will update at a later time otherwise.
Yes, indeed, the concatenation would be better before the callback.

As a side note, I haven't tested that code with camera types other than 'arcball'.

**Comment by djhoese** [comment]
I'll leave the update to you. Once that is done and we merge #1462 we can do more testing of #1463 with different cameras. I don't want to do the update because that makes it look like I understand what's going on...which I don't think I do 😉.
